src/main/java/com/linkedin/kmf/services/ConsumeMetrics.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright 2016 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.linkedin.kmf.services;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.DescribeTopicsResult;
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Percentile;
+import org.apache.kafka.common.metrics.stats.Percentiles;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.Total;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ConsumeMetrics {
+  final Sensor _consumeError;
+  final Sensor _bytesConsumed;
+  final Sensor _recordsConsumed;
+  final Sensor _recordsDuplicated;
+  final Sensor _recordsLost;
+  final Sensor _recordsDelay;
+  final Sensor _recordsDelayed;
+  private static final String METRIC_GROUP_NAME = "consume-service";
+  private static final Logger LOG = LoggerFactory.getLogger(ConsumeMetrics.class);
+
+  ConsumeMetrics(final Metrics metrics,
+      final Map<String, String> tags,
+      String topicName,
+      CompletableFuture<Void> topicPartitionReady,
+      AdminClient adminClient,
+      final int latencyPercentileMaxMs,
+      final int latencyPercentileGranularityMs) {
+    topicPartitionReady.thenRun(() -> {
+      DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(Collections.singleton(topicName));
+      Map<String, KafkaFuture<TopicDescription>> topicResultValues = describeTopicsResult.values();
+      KafkaFuture<TopicDescription> topicDescriptionKafkaFuture = topicResultValues.get(topicName);
+      TopicDescription topicDescription = null;
+      try {
+        topicDescription = topicDescriptionKafkaFuture.get();
+      } catch (InterruptedException | ExecutionException e) {
+        LOG.error("Exception occurred while retrieving the topic description.", e);
+      }
+      assert topicDescription != null;
```

Comment:
Asserts are special instructions which are not normally executed by the JVM unless the -ea flag is set when the JVM starts up.  The default is is that the flag is not set.  So asserts in Java are really only useful if the check the assert is guarding is  something expensive like an O(n) check.  For O(1) checks just throwing an exception is almost always better.

Reply:
Agreed. Revised.